docs/ARCHITECTURE.md

```diff
@@ -0,0 +1,125 @@
+# Architecture
+
+This library is under development, and it's primarily designed to fulfil needs of the
+open telemetry exporter and therefore some functionality that would be useful in other
+scenarios is not implemented.
+
+## Idea
+
+Design is heavily influenced by peculiarities of the upstream and downstream systems
+about which I know very little.
+
+### Downstream
+
+The downstream system is DataSet REST API and its API call [addEvents](https://app.scalyr.com/help/api#addEvents).
+It has the following properties:
+
+* `D1` - DataSet does not support parallel requesting per session. Only one request for each session could be made at one time
+* `D2` - when new version is deployed there is few minutes outage
```

Comment:
Could change it to something along the lines of "When a new DataSet version is deployed, there could be a short window of unavailability of a particular queue server when it's being restarted in a rolling manner. During the restart / unavailability, sessions which are mapped to a particular queue server may be temporary unavailable and those requests need to be retried. Per queue server unavailability should not last more than a couple of minutes (in real life it's < 120 seconds)".